### PR TITLE
feat: add remove notebook command

### LINKED ISSUE #32: Implement DSS `remove` command
### Why it needs to get done

The remove-notebook command removes a specified notebook using the lightkube library and prints "notebook removed" upon completion.

```
dss remove-notebook --name user-notebook
```

### What needs to get done

1. Implement `remove` command which deploys aforementioned components.
2. Implement integration test with Microk8s

### When is the task considered done

1. Command is implemented
2. Tests are passing

### LINKED ISSUE #34: Implement DSS `purge` command
### Why it needs to get done

The `remove-components` command removes all Kubernetes components connected to DSS. Internally, this involves the removal of the entire `dss` Kubernetes namespace. Importantly, this command does not uninstall the DSS snap or the Python library. It proves useful for reinitialization without the necessity to remove the snap. It's important to note that PVC for notebooks will be removed but data will stay in microk8’s [hostpath folder](https://microk8s.io/docs/addon-hostpath-storage#Customize%20directory%20used%20for%20PersistentVolume).

```
dss remove-components
```

### What needs to get done

1. Implement `remove components` command which deploys aforementioned components.
2. Implement integration test with Microk8s

### When is the task considered done

1. Command is implemented
2. Tests are passing

### PR DESCRIPTION
**Pull request by NohaIhab**
closes #32

### CODE REVIEW

File: src/dss/remove_notebook.py
Comment (ca-scribner) [inline src/dss/remove_notebook.py:24]: Just a drive-by comment on this that came up in #43 (specifically [here](https://github.com/canonical/data-science-stack/pull/43/files/627aff24391098bdb7a0506f3b11f173735eae7c#r1523105906))...  I realized create-notebook was creating its own deletion function when things timed out, which is not good.  

I wonder if we should use an `ownerReference` on the `Service` so that it will automatically get cleaned up if the notebook `Deployment` is deleted

File: src/dss/main.py
Comment (misohu) [inline src/dss/main.py:38]: I think the name of the notebook should be required. What if not provided ? Should it remove all notebooks ?

File: src/dss/remove_notebook.py
Comment (misohu) [inline src/dss/remove_notebook.py:27]: Maybe we can add some more helpful message recommendind what to do e.g. 

```
try dss list-notebooks to use the correct notebook name
```

**Review by misohu — COMMENTED** [review]
Added small comments ... still I think we should adjust the integration test and merge this after we merge @ca-scribner PR https://github.com/canonical/data-science-stack/pull/43

File: src/dss/main.py
Comment (NohaIhab) [reply src/dss/main.py:38]: good point

**Comment by misohu** [comment]
We might need to postpone this PR until this issue is resolved https://github.com/canonical/data-science-stack/issues/58

File: src/dss/main.py
Comment (NohaIhab) [reply src/dss/main.py]: done

File: src/dss/remove_notebook.py
Comment (NohaIhab) [reply src/dss/remove_notebook.py]: nice suggestion, done

**Comment by ca-scribner** [comment]
@NohaIhab I think this actually is meant to close #32, not #34.  #34 is to remove everything (all notebooks, mlflow, etc) in one shot

File: src/dss/remove_notebook.py
Comment (ca-scribner) [reply src/dss/remove_notebook.py:24]: @kimwnasptd probably has a better sense of what is appropriate.  But I wouldn't block this PR on the choice.  If this PR gets to the point where everything else is reviewed and ready to go, lets merge with whatever you choose and we can always change it after.

File: src/dss/remove_notebook.py
Comment (ca-scribner) [inline src/dss/remove_notebook.py:15]: `Notebook` is a bit abstract - can you have a second paragraph that says specifically what will be removed (the `Deployment` and `Service` of name `name`)

File: tests/integration/test_dss.py
Comment (ca-scribner) [inline tests/integration/test_dss.py:122]:
```suggestion
# FIXME: remove function and `tests/integration/deploy.yaml` when 
#  https://github.com/canonical/data-science-stack/pull/43 is merged
```

File: src/dss/remove_notebook.py
Comment (ca-scribner) [inline src/dss/remove_notebook.py:30]: Should this count as a successful execution of `dss remove-notebook`?  As is, this warns and then completes successfully, so something like:
```
dss remove-notebook notARealNotebook && echo success
```
would show the warning and then echo "success"

imo, we should fail here rather than return and succeed (`exit(1)` maybe? or probably raise a custom exception that the `main.py` catches and `main.py` then does `exit(1)`).  Its a bit arbitrary, but I don't think we want to communicate that things went as we hoped.

For comparison, doing `kubectl delete namespace not-real` exits with error code 1

**Comment by NohaIhab** [comment]
thanks @ca-scribner 
This is moving to blocked by #58, once it is unblocked we can reiterate on the comments.

**Comment by misohu** [comment]
The dss remove command has been discussed with UX team please check the DSS spec [here ](https://docs.google.com/document/d/1v0wU8TBb-ui6RcjGRbLwA6XFZnWbuiVb2TPmEgwKejY/edit#heading=h.bvvmo5r2r72m) for details about the error messages and parameters.

File: src/dss/remove_notebook.py
Comment (NohaIhab) [reply src/dss/remove_notebook.py:30]: I think this applies to all the dss commands we've implemented so far, what we've done up till now is log errors rather than exit with error code. I will bring this up with the team so we can make a decision.

File: tests/integration/test_dss.py
Comment (NohaIhab) [reply tests/integration/test_dss.py]: no need for this comment anymore

File: src/dss/remove_notebook.py
Comment (misohu) [inline src/dss/remove_notebook.py:37]: Lets rewrite the info message here the same way we o for logs https://github.com/canonical/data-science-stack/blob/main/src/dss/logs.py#L54

File: src/dss/remove_notebook.py
Comment (NohaIhab) [reply src/dss/remove_notebook.py]: added

File: src/dss/remove_notebook.py
Comment (NohaIhab) [reply src/dss/remove_notebook.py]: good point, changed now

File: tests/integration/test_dss.py
Comment (DnPlas) [inline tests/integration/test_dss.py:160]: I think you want to check that the exception that gets raised has code 404, otherwise you could be hitting false positives.

File: src/dss/remove_notebook.py
Comment (DnPlas) [reply src/dss/remove_notebook.py:26]: I believe both the labels and the `ownerReference` will be helpful here. It looks like in this PR you could consider deleting the resources with the `dss.configDSS_CLI_MANAGER_LABELS` label. Perhaps in a separate PR you could also add an `ownerReference` to the `Service` so it only gets deleted IFF the `Deployment` gets deleted, as @ca-scribner suggests.

Comment (kimwnasptd) [reply src/dss/remove_notebook.py:26]: Thanks for the ping! `ownerReferences` is something I've seen in operator pattern, in which an object owns other objects (CR owns underlying objects).

I would suggest to not go with this road for now, as this is more of an indicator we need a K8s controller. And in this case it means the CLI is this controller.

File: tests/integration/test_dss.py
Comment (NohaIhab) [reply tests/integration/test_dss.py]: agreed

File: src/dss/remove_notebook.py
Comment (DnPlas) [reply src/dss/remove_notebook.py:26]: I'm okay with what @kimwnasptd mentions, if we consider it for the future, sounds like it can be added in a separate PR. I wonder if you are still considering the labels, though @NohaIhab.

**Review by DnPlas — COMMENTED** [review]
It's looking good overall, there's still an open conversation about labels. Thanks!

File: src/dss/remove_notebook.py
Comment (ca-scribner) [inline src/dss/remove_notebook.py:26]: What about the case where only the service exists?  In that case we try to delete the Deployment, get a 404 error, and exit without trying to delete the Service. 

I propose we refactor this to ensure we try to delete every object, even if a previous one failed.  We do this same approach [in charms](https://github.com/canonical/charmed-kubeflow-chisme/blob/8a7454fede8dbcefe65cd45f2fc04a3926b7e421/src/charmed_kubeflow_chisme/lightkube/batch/_many.py#L78).

Comment (ca-scribner) [reply src/dss/remove_notebook.py:26]: current way lgtm

Comment (misohu) [reply src/dss/remove_notebook.py:26]: +1 on this

Comment (NohaIhab) [reply src/dss/remove_notebook.py:26]: @DnPlas what's your suggestion on using labels? I think what you're suggesting would apply to #75 to remove all the resources managed by DSS, but I'm not sure how it applies here. Can you elaborate?

Comment (NohaIhab) [reply src/dss/remove_notebook.py]: makes sense, I've refactored accordingly

**Review by misohu — DISMISSED** [review]
Good Job @NohaIhab

File: src/dss/remove_notebook.py
Comment (DnPlas) [reply src/dss/remove_notebook.py]: I was asking because you mentioned it in an earlier comment, but if that doesn't apply here, there is nothing else to discuss (:

**Review by misohu — APPROVED** [review]
GJ :)

**Pull request merged**
